api/analyzedphenotypes.formfields.api.inc

```diff
@@ -0,0 +1,831 @@
+    'field_hidden',  // Name attached to Name attribute of a hidden field.
+  );
+
+  // Array to hold constructed fieldnames.
+  $fieldnames = array();
+
+  // For each field, create a field name used for #id, cache, #hidden fieldname etc.
+  foreach($basenames as $base) {
+    foreach($fieldtypes as $type) {
+      $separator = ($type == 'field-id') ? '-' : '_';
+
+      // ap-fieldname-field-id.
+      $value = ($type == 'base') ? $base : 'ap' . $separator . $base . $separator . $type;
+
+      $fieldnames[$base][$type] = $value;
+    }
+  }
+
+  return $fieldnames;
+}
+
+
+///////
+
+
+/**
+ * TODO: Replace this function with the function below.
+ */
+function analyzedphenotypes_cachevalue($value) {
+  return ap_cache_download_fieldvalue($value);
+}
+
+/**
+ * FUNCTION:
+ * Form field values are save (cached) when nanipulating fields. Values are then compared
+ * to current selection to predict if new value had been selected or form retained same
+ * values in each AJAX call. The form state can then be used wheather to process a request
+ * or just reload values/selections.
+ *
+ * @param $value
+ *   Value entered or selection made to field element.
+ *
+ * @return
+ *   An array, to register a value in Drupal Form API.
+ */
+function ap_cache_download_fieldvalue($value) {
+  if (is_array($value)) {
+    $value = implode('_', $value);
+  }
+
+  return array(
+    '#type' => 'value',
+    '#value' => $value,
+  );
+}
+
+///////
+
+/**
+ * TODO: Replace this function with the function below.
+ */
+function analyzedphenotypes_autofield_callback($command, $element, $default_value = null) {
+  return ap_insert_autocomplete_formfield($command, $element, $default_value);
+}
+
+/**
+ * FUNCTION CALLBACK:
+ * AJAX add and remove form field.
+ * Dependencies: analyzedphenotypes_autofieldprop().
+ *
+ * @param $command
+ *   String, add (add) or remove (rem) to add and remove autocomplete from field element.
+ * @param $element
+ *   String, id attribute of element where to insert field in case of command add and
+ *   id attribute of field to delete in case of command rem.
+ * @param $default_value
+ *   String, default value of autocomplete field. base64 encoding used.
+ *
+ * @return
+ *   Rendered autocomplete form element.
+ *
+ * @see
+ *   hook_menu().
+ *   autocomplete form field API definition.
+ */
+function ap_insert_autocomplete_formfield($command, $element, $default_value = null) {
+  // Add or remove autofield command.
+  $cmd = trim($command);
+  $elm = trim($element);
+
+  if (($cmd == 'add' || $cmd == 'rem') && !empty($elm)) {
+    // Add or remove autofield command.
+    $element_id = '#' . $elm;
+
+    if ($cmd == 'add') {
+      // Create a from field api and append to element wrapper.
+      $default_value = base64_decode($default_value);
+      $field_api = analyzedphenotypes_autofieldprop($default_value);
+      $field_htm = drupal_render($field_api);
+
+      $AJAX_cmd = ajax_command_prepend($element_id, $field_htm);
```

Comment:
Rather then use ajax_command_prepend() and ajax_command_remove() for the germplasm selector, you should use the [typical ajax](https://www.drupal.org/docs/7/api/javascript-api/ajax-forms-in-drupal-7) just like you do for the rest of the form. The key to getting this to work in the case where fields are removed/added is to add them through the hook_form(), NOT through the AJAX callback. Here's my example of a germplasm selector although I see I don't use AJAX since I wanted the genotypes to display as germplasm were selected: https://github.com/UofS-Pulse-Binfo/nd_genotypes/blob/7.x-3.x/includes/nd_genotypes.form_elements.inc#L48

Reply:
Note: since your germplasm field is quite complicated you may want to take the route I did which involves creating a custom form elements :-). Then in your form you can easily [embed it in your form](https://github.com/UofS-Pulse-Binfo/nd_genotypes/blob/7.x-3.x/includes/nd_genotypes.genotype_matrix.inc#L61). Here's a good tutorial: https://www.silviogutierrez.com/blog/custom-drupal-elements-forms/